src/software/thunderscope/robot_diagnostics/handheld_device_manager.py

```diff
@@ -0,0 +1,341 @@
+
+    def set_enabled(self, enabled) -> None:
+        """Enable or disable the HandheldDeviceManager from processing input events.
+
+        :param enabled: whether to enable (True) or disable (False) the
+                        HandheldDeviceManager from processing input events
+        """
+        with self.lock:
+            self.enabled = enabled
+
+    def reinitialize_handheld_device(self) -> None:
+        """Attempt to reinitialize a connection to a handheld device."""
+        self.__clear_handheld_device()
+        self.__initialize_handheld_device()
+
+    def __initialize_handheld_device(self) -> None:
+        """Attempt to initialize a connection to a handheld device.
+        The first handheld device that is recognized as valid will be used.
+        Valid handheld devices are any devices whose name has a matching DeviceConfig
+        in DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP.
+        """
+        device_config_map = DiagnosticsConstants.HANDHELD_DEVICE_NAME_CONFIG_MAP
+
+        for device in list_devices():
+            handheld_device = InputDevice(device)
+            if (
+                handheld_device is not None
+                and handheld_device.name in device_config_map
+            ):
+                with self.lock:
+                    self.handheld_device = handheld_device
+                    self.handheld_device_config = device_config_map[
+                        handheld_device.name
+                    ]
+
+                    self.handheld_device_connection_status_signal.emit(
+                        HandheldDeviceConnectionChangedEvent(
+                            HandheldDeviceConnectionStatus.CONNECTED,
+                            self.handheld_device.name,
+                        )
+                    )
+
+                    self.logger.debug("Successfully initialized handheld!")
+                    self.logger.debug(
+                        'Device name: "' + self.handheld_device.name + '"'
+                    )
+                    self.logger.debug("Device path: " + self.handheld_device.path)
+
+                return
+
+        self.__clear_handheld_device()
+        self.logger.debug("Failed to initialize handheld device, check USB connection")
+        self.logger.debug("Tried the following available devices:")
+        self.logger.debug(list(map(lambda d: InputDevice(d).name, list_devices())))
+
+    def __clear_handheld_device(self) -> None:
+        """Clear the handheld device and config fields by setting them to None,
+        and emit a disconnected notification signal.
+        """
+        with self.lock:
+            self.handheld_device = None
+            self.handheld_device_config = None
+
+        self.handheld_device_connection_status_signal.emit(
+            HandheldDeviceConnectionChangedEvent(
+                HandheldDeviceConnectionStatus.DISCONNECTED, None
+            )
+        )
+
+    def __event_loop(self) -> None:
```

Comment:
Just some thoughts you could ignore and mark as resolve: 

I've been reflecting on our code structure for a while. We have several threads, each running its own event loop. The problem is that these threads don’t yield control to one another because we are running a `while True:` loop. This design doesn’t scale well with many asynchronous tasks since we only need to call `read_and_process_event` when there's an actual event to process. As a result, we waste resources and compute time by constantly checking for read events. 

We might consider using `asyncio` or a similar approach to handle these events more efficiently. This way, we can reduce the resources spent on checking for read conditions. Although, this might be a good idea, I feel likle it might but too much work for us to rewrite the full IO stack. The performance improvement may not justifiy the need.

Reply:
Async would work well in this case. But for Thunderscope in general, I'm not sure there will be a significant performance improvement from switching to async to avoid busy waiting (it actually may add overhead and slow things down) since most of our tasks are either CPU bound or just don't sit around spinning. 

e.g. in `ProtoPlayer` the playback worker thread probably spends way more time unzipping the replay chunks vs. waiting on file I/O. 

In `RobotCommunication` the thread that sends primitives pretty much always has new primitives to process, so that thread is always working (except for a bit of waiting on network I/O). 

So converting the work these threads perform into async tasks would just move a bunch of heavy processing to the Qt thread and end up blocking the UI.

Also, right now we don't have enough tasks to warrant using async for scalability benefits (we only have a single thread in `RobotCommunication` that sends primitives, unlike a web server that might handle hundreds of concurrent network requests).